`TypeError: exceptions must derive from BaseException` when unable to determine broker version

Traceback of an error when we tried upgrading to 2.1.0

Scenario, we try to connect to a kafka broker that's just starting up (so failures and retries are expected)

```
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/consumer/group.py:379: in __init__
    self._client = self.config['kafka_client'](metrics=self._metrics, **self.config)
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/client_async.py:262: in __init__
    self.config['api_version'] = self.check_version()
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/client_async.py:1052: in check_version
    self.poll(timeout_ms=timeout_ms)
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/client_async.py:683: in poll
    self._poll(timeout / 1000)
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/client_async.py:740: in _poll
    conn.connect()
/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/kafka/conn.py:436: in connect
    if self._try_api_versions_check():
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <kafka.conn.BrokerConnection object at 0x7f40eb315790>

    def _try_api_versions_check(self):
        if self._api_versions_future is None:
            if self.config['api_version'] is not None:
                self._api_version = self.config['api_version']
                self._api_versions = BROKER_API_VERSIONS[self._api_version]
                log.debug('%s: Using pre-configured api_version %s for ApiVersions', self, self._api_version)
                return True
            elif self._check_version_idx is None:
                request = ApiVersionsRequest[self._api_versions_idx]()
                future = Future()
                response = self._send(request, blocking=True, request_timeout_ms=(self.config['api_version_auto_timeout_ms'] * 0.8))
                response.add_callback(self._handle_api_versions_response, future)
                response.add_errback(self._handle_api_versions_failure, future)
                self._api_versions_future = future
                self.state = ConnectionStates.API_VERSIONS_RECV
                self.config['state_change_callback'](self.node_id, self._sock, self)
            elif self._check_version_idx < len(self.VERSION_CHECKS):
                version, request = self.VERSION_CHECKS[self._check_version_idx]
                future = Future()
                response = self._send(request, blocking=True, request_timeout_ms=(self.config['api_version_auto_timeout_ms'] * 0.8))
                response.add_callback(self._handle_check_version_response, future, version)
                response.add_errback(self._handle_check_version_failure, future)
                self._api_versions_future = future
                self.state = ConnectionStates.API_VERSIONS_RECV
                self.config['state_change_callback'](self.node_id, self._sock, self)
            else:
>               raise 'Unable to determine broker version.'
E               TypeError: exceptions must derive from BaseException
```

This screws up our retry mechanism (as we expect a ConnectionError or KafkaError, but not a type error), our current fix is to lock our usage to `2.0.6`

--- a/kafka/conn.py
+++ b/kafka/conn.py
@@ -533,7 +533,7 @@ class BrokerConnection(object):
                 self._api_version = self.config['api_version']
                 # api_version will be normalized by KafkaClient, so this should not happen
                 if self._api_version not in BROKER_API_VERSIONS:
-                    raise Errors.UnrecognizedBrokerVersion('api_version %s not found in kafka.protocol.broker_api_versions', self._api_version)
+                    raise Errors.UnrecognizedBrokerVersion('api_version %s not found in kafka.protocol.broker_api_versions' % (self._api_version,))
                 self._api_versions = BROKER_API_VERSIONS[self._api_version]
                 log.debug('%s: Using pre-configured api_version %s for ApiVersions', self, self._api_version)
                 return True

--- a/kafka/conn.py
+++ b/kafka/conn.py
@@ -531,6 +531,9 @@ class BrokerConnection(object):
         if self._api_versions_future is None:
             if self.config['api_version'] is not None:
                 self._api_version = self.config['api_version']
+                # api_version will be normalized by KafkaClient, so this should not happen
+                if self._api_version not in BROKER_API_VERSIONS:
+                    raise Errors.UnrecognizedBrokerVersion('api_version %s not found in kafka.protocol.broker_api_versions', self._api_version)
                 self._api_versions = BROKER_API_VERSIONS[self._api_version]
                 log.debug('%s: Using pre-configured api_version %s for ApiVersions', self, self._api_version)
                 return True
@@ -553,7 +556,8 @@ class BrokerConnection(object):
                 self.state = ConnectionStates.API_VERSIONS_RECV
                 self.config['state_change_callback'](self.node_id, self._sock, self)
             else:
-                raise 'Unable to determine broker version.'
+                self.close(Errors.KafkaConnectionError('Unable to determine broker version.'))
+                return False
 
         for r, f in self.recv():
             f.success(r)

--- a/test/test_conn.py
+++ b/test/test_conn.py
@@ -93,6 +93,12 @@ def test_api_versions_check(_socket):
     assert conn.disconnected() is True
 
 
+def test_api_versions_check_unrecognized(_socket):
+    conn = BrokerConnection('localhost', 9092, socket.AF_INET, api_version=(0, 0))
+    with pytest.raises(Errors.UnrecognizedBrokerVersion):
+        conn.connect()
+
+
 def test_connect_timeout(_socket, conn):
     assert conn.state is ConnectionStates.DISCONNECTED
 

--- a/test/test_conn.py
+++ b/test/test_conn.py
@@ -69,6 +69,30 @@ def test_connect(_socket, conn, states):
         assert conn.state is state
 
 
+def test_api_versions_check(_socket):
+    conn = BrokerConnection('localhost', 9092, socket.AF_INET)
+    assert conn._api_versions_future is None
+    conn.connect()
+    assert conn._api_versions_future is not None
+    assert conn.connecting() is True
+    assert conn.state is ConnectionStates.API_VERSIONS_RECV
+
+    assert conn._try_api_versions_check() is False
+    assert conn.connecting() is True
+    assert conn.state is ConnectionStates.API_VERSIONS_RECV
+
+    conn._api_versions_future = None
+    conn._check_version_idx = 0
+    assert conn._try_api_versions_check() is False
+    assert conn.connecting() is True
+
+    conn._check_version_idx = len(conn.VERSION_CHECKS)
+    conn._api_versions_future = None
+    assert conn._try_api_versions_check() is False
+    assert conn.connecting() is False
+    assert conn.disconnected() is True
+
+
 def test_connect_timeout(_socket, conn):
     assert conn.state is ConnectionStates.DISCONNECTED
 

